dtfs-central-api/src/v1/controllers/portal/deal/update-deal-status.controller.js

```diff
@@ -1,58 +1,67 @@
+const { validateAuditDetails, generateAuditDatabaseRecordFromAuditDetails } = require('@ukef/dtfs2-common/change-stream');
+const { MONGO_DB_COLLECTIONS, InvalidAuditDetailsError } = require('@ukef/dtfs2-common');
 const { ObjectId } = require('mongodb');
 const $ = require('mongo-dot-notation');
 const { findOneDeal } = require('./get-deal.controller');
 const { mongoDbClient: db } = require('../../../../drivers/db-client');
 
-const withoutId = (obj) => {
-  const cleanedObject = { ...obj };
-  delete cleanedObject._id;
-  return cleanedObject;
-};
-
-const updateDealStatus = async (dealId, status, existingDeal) => {
-  if (ObjectId.isValid(dealId)) {
-    const dealsCollection = await db.getCollection(MONGO_DB_COLLECTIONS.DEALS);
-
-    const previousStatus = existingDeal.status;
+const updateDealStatus = async ({ dealId, status, existingDeal, auditDetails }) => {
+  const dealsCollection = await db.getCollection(MONGO_DB_COLLECTIONS.DEALS);
 
-    const modifiedDeal = {
-      ...existingDeal,
-      updatedAt: Date.now(),
-      status,
-      previousStatus,
-    };
+  const previousStatus = existingDeal.status;
 
-    const findAndUpdateResponse = await dealsCollection.findOneAndUpdate({ _id: { $eq: ObjectId(dealId) } }, $.flatten(withoutId(modifiedDeal)), {
+  const modifiedDeal = {
+    ...existingDeal,
+    updatedAt: Date.now(),
+    status,
+    previousStatus,
+  };
+  const auditRecord = generateAuditDatabaseRecordFromAuditDetails(auditDetails);
+  const { _id, ...modifiedDealWithoutId } = modifiedDeal;
+  const findAndUpdateResponse = await dealsCollection.findOneAndUpdate(
+    { _id: { $eq: ObjectId(dealId) } },
+    $.flatten({ ...modifiedDealWithoutId, auditRecord }),
+    {
       returnNewDocument: true,
       returnDocument: 'after',
-    });
+    },
+  );
 
-    console.info('Updated Portal BSS deal status from %s to %s', previousStatus, status);
+  console.info('Updated Portal BSS deal status from %s to %s', previousStatus, status);
 
-    return findAndUpdateResponse.value;
-  }
-  return { status: 400, message: 'Invalid Deal Id' };
+  return findAndUpdateResponse.value;
 };
-exports.updateDealStatus = updateDealStatus;
 
 exports.updateDealStatusPut = async (req, res) => {
-  if (ObjectId.isValid(req.params.id)) {
-    const dealId = req.params.id;
-
-    const { status } = req.body;
-
-    return await findOneDeal(dealId, async (existingDeal) => {
-      if (existingDeal) {
-        if (existingDeal.status === status) {
-          return res.status(400).send();
-        }
-        const updatedDeal = await updateDealStatus(dealId, status, existingDeal);
-        return res.status(200).json(updatedDeal);
-      }
-      return res.status(404).send({ status: 404, message: 'Deal not found' });
-    });
+  const {
+    params: { id: dealId },
+    body: { status, auditDetails },
+  } = req;
+
+  if (!ObjectId.isValid(dealId)) {
+    return res.status(400).send({ status: 400, message: 'Invalid Deal Id' });
+  }
+
+  try {
```

Comment:
Not saying to do this here at all*
how long do you think it would take to turn this into middleware that just runs this code? Is it just probably copy pasting a middleware onto every route that has this then deleeting this everywhere?

Reply:
I really think it would be nice to
1. create audit details on entering portal-api/tfm-api in middleware
2. create a validate audit details/transform into audit record in dtfs-central-api
Then it should just be a job of passing things around.

Realistically, I would put bullet 2 at around 5 points, but I wouldn't be surprised if it were a 3.
I think bullet 1 above at around the same, as we should already be constructing these at `res/req` functions but this is not always the case (sometimes we construct the audit details in tfm-api after receiving a request from portal-api, which im not keen on)